esphome/components/kamstrup_kmp/kamstrup_kmp.h

```diff
@@ -0,0 +1,128 @@
+    and by measuring the water flow. The temperature difference (delta T)
+    together with the water flow results in consumed energy, typically
+    in giga joule (GJ).
+
+    The Kamstrup Multical has an optical interface just above the display.
+    This interface is essentially an RS-232 interface using a proprietary
+    protocol (Kamstrup Meter Protocol [KMP]).
+
+    The integration uses this optical interface to periodically read the
+    configured values (sensors) from the meter. Supported sensors are:
+      - Heat Energy               [GJ]
+      - Current Power Consumption [kW]
+      - Temperature 1             [°C]
+      - Temperature 2             [°C]
+      - Temperature Difference    [°K]
+      - Water Flow                [l/h]
+      - Volume                    [m3]
+
+    Apart from these supported 'fixed' sensors, the user can configure up to
+    five custom sensors. The KMP command (16 bit unsigned int) has to be
+    provided in that case.
+
+    Note:
+    The optical interface is enabled as soon as a button on the meter is pushed.
+    The interface stays active for a few minutes. To keep the interface 'alive'
+    magnets must be placed around the optical sensor.
+
+    Units:
+    Units are set using the regular Sensor config in the user yaml. However,
+    KMP does also send the correct unit with every value. When DEBUG logging
+    is enabled, the received value with the received unit are logged.
+
+    Acknowledgement:
+    This interface was inspired by:
+      - https://atomstar.tweakblogs.net/blog/19110/reading-out-kamstrup-multical-402-403-with-home-built-optical-head
+      - https://wiki.hal9k.dk/projects/kamstrup
+*/
+
+// KMP Commands
+static const uint16_t CMD_HEAT_ENERGY = 0x003C;
+static const uint16_t CMD_POWER = 0x0050;
+static const uint16_t CMD_TEMP1 = 0x0056;
+static const uint16_t CMD_TEMP2 = 0x0057;
+static const uint16_t CMD_TEMP_DIFF = 0x0059;
+static const uint16_t CMD_FLOW = 0x004A;
+static const uint16_t CMD_VOLUME = 0x0044;
+
+// KMP units
+static const char *const UNITS[] = {
+    "",      "Wh",   "kWh",  "MWh",   "GWh",     "J",       "kJ",       "MJ",       "GJ",       "Cal",
+    "kCal",  "Mcal", "Gcal", "varh",  "kvarh",   "Mvarh",   "Gvarh",    "VAh",      "kVAh",     "MVAh",
+    "GVAh",  "kW",   "kW",   "MW",    "GW",      "kvar",    "kvar",     "Mvar",     "Gvar",     "VA",
+    "kVA",   "MVA",  "GVA",  "V",     "A",       "kV",      "kA",       "C",        "K",        "l",
+    "m3",    "l/h",  "m3/h", "m3xC",  "ton",     "ton/h",   "h",        "hh:mm:ss", "yy:mm:dd", "yyyy:mm:dd",
+    "mm:dd", "",     "bar",  "RTC",   "ASCII",   "m3 x 10", "ton x 10", "GJ x 10",  "minutes",  "Bitfield",
+    "s",     "ms",   "days", "RTC-Q", "Datetime"};
+
+class KamstrupKMPComponent : public PollingComponent, public uart::UARTDevice {
+ public:
+  void set_heat_energy_sensor(sensor::Sensor *sensor) { heat_energy_sensor_ = sensor; }
```

Comment:
Please make sure to prefix all member variables and function calls with `this->`
```suggestion
  void set_heat_energy_sensor(sensor::Sensor *sensor) { this->heat_energy_sensor_ = sensor; }
```

Reply:
Added all missing `this->`. 
Fully tested again on the actual hardware.